### PR TITLE
Fix(Humanoid): Prevent skin color verification failures due to precision loss

### LINKED ISSUE #40832: Skin tones created by `ISkinColorationStrategy.ClosestSkinColor` often fail `ISkinColorationStrategy.VerifySkinColor`
### What version did the issue occur in?

ef0c9ecc6350d4d6bcf9c8ad9013724688dbcae3

### Description

Due to float precision and HSL/HSV conversion, skin tones created by `ClampedHslColoration.ClosestSkinColor` and `ClampedHsvColoration.ClosestSkinColor` often fail their own verification via `ClampedHslColoration.VerifySkinColor` and `ClampedHsvColoration.VerifySkinColor`.

This is causing tests to fail in #36493 as I am trying to `MemberwiseEquals` character profiles before and after server character selection, but the comparisons fail due to minute float precision differences.

### Reproduction

Apply this patch to add a test and run it.

```
Index: Content.Tests/Shared/Preferences/Humanoid/SkinTonesTest.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Content.Tests/Shared/Preferences/Humanoid/SkinTonesTest.cs b/Content.Tests/Shared/Preferences/Humanoid/SkinTonesTest.cs
--- a/Content.Tests/Shared/Preferences/Humanoid/SkinTonesTest.cs	(revision eaf3c0bd3ab712e2e3ae6ed4836faec08f2a5807)
+++ b/Content.Tests/Shared/Preferences/Humanoid/SkinTonesTest.cs	(date 1760206839479)
@@ -1,5 +1,7 @@
 using Content.Shared.Humanoid;
 using NUnit.Framework;
+using Robust.Shared.Maths;
+using Robust.Shared.Random;
 
 namespace Content.Tests.Shared.Preferences.Humanoid;
 
@@ -25,4 +27,22 @@
 
         Assert.That(strategy.VerifySkinColor(strategy.ValidHumanSkinTone));
     }
+
+    [Test]
+    public void TestTintedHuesValid()
+    {
+        var random = new RobustRandom() as IRobustRandom;
+        var strategy = new ClampedHslColoration()
+        {
+            Saturation = (0.0f, 0.1f),
+            Lightness = (0.85f, 1.0f),
+        };
+
+        for (var i = 0; i <= 10000; i++)
+        {
+            var color = new Color(random.NextFloat(1f), random.NextFloat(1f), random.NextFloat(1f));
+            var skinColor = strategy.ClosestSkinColor(color);
+            Assert.That(strategy.VerifySkinColor(skinColor));
+        }
+    }
 }
```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Uberration**
## About the PR
This PR resolves a fundamental precision issue within the `ISkinColorationStrategy` prototypes, which caused `VerifySkinColor` to fail on valid, programmatically generated colors. By correcting the backend logic, it fixes the bug described in #40832.

## Why / Balance
The primary issue is that `ClosestSkinColor` would produce colors that `VerifySkinColor` would then reject due to minuscule floating-point drift. This is a correctness issue in the prototype's implementation and can lead to unexpected behavior in systems that rely on this verification, such as UI elements or profile serialization.

**Fixes:**
- Resolves #40832 - Skin color verification failures due to precision loss

**Related Issues:**
While this fix is foundational, it may be related to but does not fully resolve:
- #40770 - Human skin tones jumping around when changing colors

Related
- #39175 / #36493

## Technical details
The root cause was floating-point drift from the `RGB -> HSL/HSV -> Clamp -> RGB` conversion cycle. The previous implementation had a hardcoded epsilon that was too small and contained complex, multi-stage logic in `ClosestSkinColor` as a workaround.

This PR introduces a fix based on diagnostic testing:

1.  A diagnostic test was created to measure the raw precision loss over millions of random color conversions (50M+ iterations).
2.  The maximum observed drift was determined to be `~4.9E-6`.
3.  The `Epsilon` constant in `ClampedHslColoration` and `ClampedHsvColoration` has been updated to `1e-5f`, providing a robust 2x safety margin over this worst-case drift.
4.  With the correctly tuned epsilon, the complex workaround logic in `ClosestSkinColor` was removed and replaced with a simple, direct clamp, improving code clarity and maintainability.

## Media
N/A

## Requirements
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes
**None** - All changes are internal improvements to prototype logic that maintain existing APIs and functionality.

**Changelog**
:cl: 
- fix: Prevent skin color verification failures due to precision loss.

### CODE REVIEW

**Comment by Uberration** [comment]
Code is rather monolithic, refactoring would be nice.

**Comment by Uberration** [comment]
Tested UI elements, they all seem to be working, dotnet test was successful locally
H 360 still rolls over to 0, when saved, but I believe that is intention.

**Comment by Uberration** [comment]
Might be better to move the refactor to another PR

**Comment by sleepyyapril** [comment]
id still recommend using UI elements over just switching to using managers, because frankly that'd be far easier for downstreams to use.
image
for example would be FAR easier to maintain than keeping HumanoidProfileEditor as the one handling initialization

File: Content.Shared/Humanoid/SkinColorationPrototype.cs [hunk 1]
```diff
@@ -182,43 +183,36 @@ public float ToUnary(Color color)
...
-    public (float, float)? Saturation;
-
-    /// <summary>
-    /// The (min, max) of the value channel.
-    /// </summary>
-    [DataField]
-    public (float, float)? Value;
+    [DataField] public (float, float)? Hue;
+    [DataField] public (float, float)? Saturation;
+    [DataField] public (float, float)? Value;
```
Comment (slarticodefast) [inline Content.Shared/Humanoid/SkinColorationPrototype.cs]: don't remove the documentation and keep the attribute in a separate line

File: Content.Shared/Humanoid/SkinColorationPrototype.cs [hunk 2]
```diff
@@ -289,14 +299,34 @@ public Color ClosestSkinColor(Color color)
...
 
-        return Color.FromHsl(hsl);
+    private static float ClampHue(float hue, float minHue, float maxHue)
+    {
+        if (minHue > maxHue) // Wraps around 1.0
+        {
+            if (hue >= minHue || hue <= maxHue)
+                return hue;
+
+            var mid = (minHue + maxHue + 1) / 2f;
```
Comment (slarticodefast) [inline Content.Shared/Humanoid/SkinColorationPrototype.cs]:
```suggestion
            var mid = (minHue + maxHue) / 2f;
```
shouldn't the midpoint be this?

File: Content.Tests/Shared/Preferences/Humanoid/SkinTonesTest.cs [hunk 3]
```diff
@@ -1,28 +1,263 @@
...
+        TestContext.Out.WriteLine("\n--- FINAL DRIFT SUMMARY FOR ALL CLAMPING TESTS ---");
+        TestContext.Out.WriteLine($"Maximum observed HSL drift: {_maxHslDrift:E}"); // Scientific notation for precision
+        TestContext.Out.WriteLine($"Maximum observed HSV drift: {_maxHsvDrift:E}");
+        TestContext.Out.WriteLine("This indicates the actual max floating-point error observed. Production code's Epsilon should be >= this value.");
+        TestContext.Out.WriteLine("--------------------------------------------------");
+    }
+
     [Test]
-    public void TestHumanSkinToneValidity()
+    public void TestHumanSkinTonesFromUnaryAreValid()
```
Comment (slarticodefast) [inline Content.Tests/Shared/Preferences/Humanoid/SkinTonesTest.cs:37]: add documentation for each test

File: Content.Tests/Shared/Preferences/Humanoid/SkinTonesTest.cs [hunk 4]
```diff
@@ -1,28 +1,263 @@
...
     {
         var strategy = new HumanTonedSkinColoration();
 
-        for (var i = 0; i <= 100; i++)
+        // Testing across a finer range to hit more edge cases
+        for (var i = 0; i <= 10000; i++)
```
Comment (slarticodefast) [inline Content.Tests/Shared/Preferences/Humanoid/SkinTonesTest.cs:42]: Probably not an issue, but have you checked how long these tests run if we do 10k steps for each test?

Comment (Uberration) [reply Content.Tests/Shared/Preferences/Humanoid/SkinTonesTest.cs:42]: > Probably not an issue, but have you checked how long these tests run if we do 10k steps for each test?

succeeded (0.8s) on my own system, but might be slightly longer older hardware.

File: Content.Shared/Humanoid/SkinColorationPrototype.cs [hunk 5]
```diff
@@ -229,25 +237,55 @@ public Color ClosestSkinColor(Color color)
...
+    private static bool IsHueInRange(float hue, float minHue, float maxHue)
+    {
+        if (minHue > maxHue) // Wraps around 1.0 (e.g., reds)
+            return hue >= minHue - Epsilon || hue <= maxHue + Epsilon;
+        return hue >= minHue - Epsilon && hue <= maxHue + Epsilon;
+    }
 
-        if (Value is (var minValue, var maxValue))
-            hsv.Z = Math.Clamp(hsv.Z, minValue, maxValue);
+    private static float ClampHue(float hue, float minHue, float maxHue)
```
Comment (sleepyyapril) [inline Content.Shared/Humanoid/SkinColorationPrototype.cs]: should not be making this twice

File: Content.Shared/Humanoid/SkinColorationPrototype.cs [hunk 6]
```diff
@@ -289,14 +327,35 @@ public Color ClosestSkinColor(Color color)
...
-            hsl.Z = Math.Clamp(hsl.Z, minValue, maxValue);
+    private static bool IsHueInRange(float hue, float minHue, float maxHue)
+    {
+        if (minHue > maxHue) // Wraps around 1.0 (e.g., reds)
+            return hue >= minHue - Epsilon || hue <= maxHue + Epsilon;
+        return hue >= minHue - Epsilon && hue <= maxHue + Epsilon;
+    }
 
-        return Color.FromHsl(hsl);
+    private static float ClampHue(float hue, float minHue, float maxHue)
```
Comment (sleepyyapril) [inline Content.Shared/Humanoid/SkinColorationPrototype.cs]: same

File: Content.Shared/Humanoid/SkinColorationPrototype.cs [hunk 7]
```diff
@@ -229,25 +237,55 @@ public Color ClosestSkinColor(Color color)
...
+            hsv.Y = Math.Clamp(hsv.Y, minSat, maxSat);
+        if (Value is (var minVal, var maxVal))
+            hsv.Z = Math.Clamp(hsv.Z, minVal, maxVal);
+
+        return Color.FromHsv(hsv);
+    }
 
-        if (Saturation is (var minSaturation, var maxSaturation))
-            hsv.Y = Math.Clamp(hsv.Y, minSaturation, maxSaturation);
+    private static bool IsHueInRange(float hue, float minHue, float maxHue)
```
Comment (sleepyyapril) [inline Content.Shared/Humanoid/SkinColorationPrototype.cs]: should not be making the same function twice

File: Content.Shared/Humanoid/SkinColorationPrototype.cs [hunk 8]
```diff
@@ -289,14 +327,35 @@ public Color ClosestSkinColor(Color color)
...
+            hsl.Z = Math.Clamp(hsl.Z, minLight, maxLight);
 
-        if (Saturation is (var minSaturation, var maxSaturation))
-            hsl.Y = Math.Clamp(hsl.Y, minSaturation, maxSaturation);
+        return Color.FromHsl(hsl);
+    }
 
-        if (Lightness is (var minValue, var maxValue))
-            hsl.Z = Math.Clamp(hsl.Z, minValue, maxValue);
+    private static bool IsHueInRange(float hue, float minHue, float maxHue)
```
Comment (sleepyyapril) [inline Content.Shared/Humanoid/SkinColorationPrototype.cs]: same

Comment (Uberration) [reply Content.Shared/Humanoid/SkinColorationPrototype.cs]: moved to internal helper class [9f2ae76](https://github.com/space-wizards/space-station-14/pull/41054/commits/9f2ae76535a405fba52c563566aacf8736136319)
